Create fabric-brewing-v1 module

swordglowsblue opened the pull request:
Vanilla currently has no way to add new brewing stand recipes; all brewing recipes are hardcoded in `BrewingRecipeRegistry`. This API module extends the vanilla brewing stand with JSON recipe compatibility.

As of the initial PR, this module *does not* modify or disable any of the default vanilla recipes. It simply injects its own crafting logic ahead of vanilla's, deferring to the existing implementation if no valid custom recipes are found. If making vanilla brewing recipes data-driven is determined to be within the scope of the Fabric API, I would be happy to change this.

To create a custom brewing recipe, the following JSON recipe format can be used:

```json
{
  "type": "fabric:brewing",
  "input": { "item": "minecraft:stick" },
  "base": { "potion": "minecraft:leaping", "type": "splash" },
  "output": { "potion": "minecraft:strong_swiftness" }
}
```

`input` is the item in the top slot of the brewing stand, `base` is the item in any of the three bottom slots, and `output` is the resulting item. All three can use the standard `item` ingredient format, and `input` and `base` can use the `tag` ingredient format. Additionally, all three can use the special `PotionIngredient` format, consisting of the keys `potion` (the potion identifier) and `type` (`normal`, `splash`, or `lingering` - defaults to `normal`).  A list of vanilla potion identifiers can be found in `net.minecraft.potion.Potions`, and they generally consist of a base identifier preceded by `long_` or `strong_` if applicable. This should be fully compatible with modded potions as well.

Potential considerations / additions:

- Should vanilla's recipes be overridden and converted to JSON recipes (currently not)?
  - If so, how should brewing recipe mixins by existing mods be handled?
- Should the recipe type be namespaced under `fabric` or `minecraft` (currently `fabric`)?
- `output` does not currently support the `count` property that most recipes have.
- Recipe remainders for `input` are supported, but currently cannot be extracted via hopper.
- Custom brewing fuel types are not currently supported.
- Are API interfaces for `BrewingRecipe`/`PotionIngredient` needed?

asiekierka commented:
Generally speaking, we have discussed this before (wrt NBT crafting) come to the conclusion that custom JSON recipe formats fall outside of "hooks and mod interop", and Fabric should just provide thin hooks for the relevant areas instead. We might want to discuss this again, however.

swordglowsblue commented:
There really isn't a good way to put a thin layer over Vanilla's brewing recipe system, unfortunately. While possible, in my experience it involves several layers of mixins and lambda passing that quickly becomes quite messy. This is much more robust and extendable, as well as very lightweight/vanilla-like compared to existing third-party solutions (i.e. [NbtCrafting](https://github.com/Siphalor/nbt-crafting)). I completely understand if it's too far out of scope, though.

liach commented on fabric-brewing-v1/src/main/java/net/fabricmc/fabric/impl/brewing/PotionIngredient.java:
```diff
@@ -0,0 +1,110 @@
+    public enum Type {
+        NORMAL(Items.POTION),
+        SPLASH(Items.SPLASH_POTION),
+        LINGERING(Items.LINGERING_POTION);
```
Don't use an enum in an API when you or another mod may add more entries to it in the future.

swordglowsblue replied:
Noted, I'll see about changing that.

liach commented on fabric-brewing-v1/src/main/java/net/fabricmc/fabric/impl/brewing/BrewingRecipe.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    public static List<BrewingRecipe> getRelevantRecipes(Inventory inv, ItemStack stack, int slot) {
+        World world = inv instanceof BrewingStandBlockEntity
```
Consider modded inventories.

swordglowsblue replied:
What would you suggest as a better way to manage this? `instanceof BlockEntity`? 

Also, this method is primarily a utility for working around the existing vanilla code, and shouldn't generally need to be called from a custom block entity that doesn't have to work around that.

liach replied:
I suggest another method that allows mods to pass in a custom world instance instead of guessing wildly.

swordglowsblue replied:
Alright, I'll add a version that includes a `world` parameter as well.

liach commented on fabric-brewing-v1/src/main/java/net/fabricmc/fabric/impl/brewing/PotionIngredient.java:
```diff
@@ -0,0 +1,110 @@
+import net.minecraft.util.PacketByteBuf;
+import net.minecraft.util.registry.Registry;
+
+public class PotionIngredient {
```
Should be a custom ingredient entry instead of a new class which no one knows if this is actually backed by a potion+type or a vanilla ingredient

swordglowsblue replied:
The `Ingredient` class is unfortunately `final`, or I would have extended it directly as I'd much prefer to. How would you suggest implementing this better while still fully supporting `Ingredient` as a valid value?

liach replied:
Ah, keep this one. My bad! Thought Ingredient Entry was public. Time for me to consider exposing it!

swordglowsblue replied:
Ah, that's what you meant. `Ingredient.Entry` is public, but there isn't a way to override the methods on `Ingredient` for parsing/creating `Entry`s without a mixin, so it would be quite the process to wrap in that way. I'll see if it's feasible, though.

liach replied:
I suggest leaving that for this pr. Another one that correctly fixes this issue can come back to this later and serialize things properly instead.

Prospector commented:
Personally, I support JSON recipe formats in Fabric API, perhaps with a `fabric-json-data-extensions` module. I think it's something Mojang will do fairly soon and it would be stupid to have 100 mods add their own recipe formats for this when fabric could just add one. Just think of the nightmare when a load of different mods add their JSON recipe formats that all have slightly different field names or layout. That doesn't sound good for anyone making a data pack at all.

Pyrofab commented:
I am not particularly against JSON recipes in Fabric API, but nobody said 100 mods should have their own recipe format. A library could very well offer this feature like cotton does for cauldron recipes.

Prospector commented:
For niche features, sure. But no, I'm not gonna depend on an extra lib for recipes.

swordglowsblue commented:
In my personal experience, library mods tend to go unnoticed unless they're picked up by an extremely large mod. That's better than nothing, certainly, but having a single robust implementation baked into the "default" API will mean better adoption and less hassle for both users and developers.

swordglowsblue commented:
Some changes:

- Refactored to an API/Impl code style like other API modules.
- Changed `BrewingRecipe.getRelevantRecipes` to `BrewingRecipes.INSTANCE.getRelevantRecipes`, now takes a `World` as its first parameter instead of an `Inventory`.
- Removed the `PotionIngredient.Type` enum in favor of an `Identifier`-based `PotionTypeRegistry`.

asiekierka commented on fabric-brewing-v1/src/main/java/net/fabricmc/fabric/api/brewing/BrewingRecipe.java:
```diff
@@ -0,0 +1,29 @@
+	PotionIngredient getInput();
+	PotionIngredient getBasePotion();
+	PotionIngredient getOutputPotion();
+	List<Integer> getMatchedSlots();
```
Why not an IntList from fastutil?

asiekierka commented on fabric-brewing-v1/src/main/java/net/fabricmc/fabric/api/brewing/PotionIngredient.java:
```diff
@@ -0,0 +1,63 @@
+	Potion asPotion();
+	Identifier getPotionType();
+	boolean test(ItemStack stack);
+	void write(PacketByteBuf buf);
```
So much to implement. This is depressing to me - it's a lot to implement for custom code. Just like BrewingRecipe. It feels as if it were designed with the JSON files in mind and then the API-able components were just moved to api/. I don't think that's good to encourage.

swordglowsblue replied:
I'll admit - that's because it was. This isn't really the type of thing that needs much of an API layer, but it was my first PR to Fabric and I was a little over-concerned with matching the architecture of existing modules. I should probably go through again and clean this up more, or just drop it since it's a controversial addition anyways.

Player3324 commented:
Please rebase to 1.15 and make sure it builds (->checkstyle) before we can start reviewing and integrating this.

Prospector commented:
@swordglowsblue Do you plan on updating this PR or should it be closed?

kgscialdone commented:
I'm currently not working on any Minecraft-related projects. Feel free to close it, I can always reopen or push a new one if I ever come back to it.
